src/apprt/gtk/GlobalShortcuts.zig

```diff
@@ -0,0 +1,391 @@
+    // ask clients to manually construct the expected object path and subscribe
+    // to the request signal beforehand, making the whole response value of
+    // the original call COMPLETELY MEANINGLESS.
+    //
+    // Furthermore, this is _entirely undocumented_ aside from one tiny
+    // paragraph under the documentation for the Request interface, and
+    // anyone would be forgiven for missing it without reading the libportal
+    // source code.
+    //
+    // When in Rome, do as the Romans do, I guess...?
+
+    const callbacks = struct {
+        fn gotResponseHandle(
+            source: ?*gobject.Object,
+            res: *gio.AsyncResult,
+            _: ?*anyopaque,
+        ) callconv(.c) void {
+            const dbus_ = gobject.ext.cast(gio.DBusConnection, source.?).?;
+
+            var err: ?*glib.Error = null;
+            defer if (err) |err_| err_.free();
+
+            const params_ = dbus_.callFinish(res, &err) orelse {
+                if (err) |err_| log.err("request failed={s} ({})", .{
+                    err_.f_message orelse "(unknown)",
+                    err_.f_code,
+                });
+                return;
+            };
+            defer params_.unref();
+
+            // TODO: XDG recommends updating the signal subscription if the actual
+            // returned request path is not the same as the expected request
+            // path, to retain compatibility with older versions of XDG portals.
+            // Although it suffers from the race condition outlined above,
+            // we should still implement this at some point.
+        }
+
+        // See https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.Request.html#org-freedesktop-portal-request-response
+        fn responded(
+            _: *gio.DBusConnection,
+            _: ?[*:0]const u8,
+            _: [*:0]const u8,
+            _: [*:0]const u8,
+            _: [*:0]const u8,
+            params_: *glib.Variant,
+            ud: ?*anyopaque,
+        ) callconv(.c) void {
+            // TODO: Unsubscribe from the signal as responses are one-shot
+            var response: u32 = 0;
+            var vardict: ?*glib.Variant = null;
+            params_.get("(u@a{sv})", &response, &vardict);
+
+            switch (response) {
+                0 => log.debug("request successful", .{}),
+                1 => log.debug("request was cancelled by user", .{}),
+                2 => log.warn("request ended unexpectedly", .{}),
+                else => log.err("unrecognized response code={}", .{response}),
+            }
+
+            method.onResponse(@ptrCast(@alignCast(ud)), vardict.?);
+        }
+    };
+
+    var request_token_buf: Token = undefined;
+    const request_token = generateToken(&request_token_buf);
+
+    const payload = method.makePayload(self, request_token) orelse return;
+    const request_path = try self.getRequestPath(request_token);
+
+    _ = self.dbus.signalSubscribe(
```

Comment:
Is it a leak if we don't unsubscribe at some point?

Reply:
I'm honestly not sure, it's *recommended* to unsubscribe when you're done with it but I think not even libportal is truly strict about it

Reply:
Ah, looks like their idiomatic pattern is to store all signal IDs in an array and then removing them in one batch. Honestly that's pretty smart - I might steal that

Reply:
Yeah, I confirmed this is a memory leak thanks to Georges from Igalia!

https://gitlab.gnome.org/GNOME/glib/-/blob/main/gio/gdbusconnection.c?ref_type=heads#L3916-3922